### PR TITLE
iAPI Router: Handle styles assets on region-based navigation

### PR DESCRIPTION
**Pull request by DAreRodz**
## What?

Adds support for style assets on region-based navigation, allowing stylesheets to be applied or removed when navigating to a new page with different styles.

## Why?

This feature is required for some core blocks like the Comments one. In this particular case, when publishing a new comment on a post that does not have any comments, we want to display the new comments without reloading the page using `navigate()`. The new comment might require styles not present on the initial post page.

## How?

The implementation uses the [`adoptedStyleSheets` API](https://developer.mozilla.org/en-US/docs/Web/API/Document/adoptedStyleSheets).

For each `<style>` and `<link rel=stylesheet>` tags, a `CSSStyleSheet` is generated. If the CSS code contains a relative URL pointing to another asset (e.g., images or fonts), the URL is converted into an absolute URL. That could be removed once the `baseURL` option is supported in Chrome, Edge, or Opera (see [coverage](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet#:~:text=options.baseURL%20parameter) in MDN).

These `CSSStyleSheet` instances are then passed to the `adoptedStyleSheets` property in the correct order.

With these APIs, we can solve the following problems:
- Wait for all the styles to be ready before updating the HTML.
- Update both the styles and the HTML at the same time.
- Reuse all previous stylesheets without having to parse the CSS code again.

The PR also replaces the current implementation we have for full-site navigation.

## Testing Instructions

1. In the site editor, go to the home template
2. Ensure the "Force page reload" setting is disabled in the Query block.
3. Add an image block inside the Post Template.
4. Change its style and make the image rounded.
5. Visit a page of the home that doesn't exist (e.g., page 2) so it shows the "No Results" block.
6. Click on the Previous Page link.
7. Images should appear rounded.

**Important:** you could notice elements outside the Query block that their appearance change. It's related to [this issue](https://github.com/WordPress/gutenberg/issues/67308), and will be addressed in a subsequent PR.

### CODE REVIEW

File: packages/interactivity-router/src/index.ts [hunk 1]
```diff
@@ -169,23 +189,14 @@ window.addEventListener( 'popstate', async () => {
...
-		await fetchHeadAssets( document );
-	}
-}
 pages.set(
 	getPagePath( window.location.href ),
-	Promise.resolve( regionsToVdom( document, { vdom: initialVdom } ) )
+	Promise.resolve(
+		regionsToVdom( document, {
+			vdom: initialVdom,
+			baseUrl: window.location.href,
```
Comment (michalczaplinski) [inline packages/interactivity-router/src/index.ts:197]: Should this `baseURL` be the same as the "canonical URL" [here](https://github.com/WordPress/gutenberg/pull/67289/files#diff-ef7ce82e804956a7d5c001fefbceabf1820872331c6282c9d359cac1e7390734R217-R227) ? 🤔 By that I mean that we can set it on the server and pass it to the client like I do in my PR.

In my PR, we need the "canonical URL" which is the URL of the post/page/template without the pagination parameter (`/page/123`, etc.).

File: packages/interactivity-router/src/index.ts [hunk 2]
```diff
@@ -103,15 +110,28 @@ const regionsToVdom: RegionsToVdom = async ( dom, { vdom } = {} ) => {
 	}
 	const title = dom.querySelector( 'title' )?.innerText;
 	const initialData = parseServerData( dom );
-	return { regions, head, title, initialData };
+	return { regions, styles, scriptModules, title, initialData };
 };
 
 // Render all interactive regions contained in the given page.
 const renderRegions = async ( page: Page ) => {
+	// Whait for styles and modules to be ready.
```
Comment (michalczaplinski) [inline packages/interactivity-router/src/index.ts]: Whait --> Wait

```suggestion
	// Wait for styles and modules to be ready.
```

**Review by michalczaplinski — COMMENTED** [review]
Gave it an initial review and overall the idea of using the `adoptedStyleSheets` is looking good to me. 👍 

I have one concern though: In this PR, we're effectively changing the behavior of prefetching. Previously, to "prefetch" meant to fetch the HTML **_as well as the JS and CSS assets of a page_**. 

Now, with the changes introduced here, prefetching **_only fetches the HTML_** of a new page and fetching of JS and CSS assets is delayed until a navigation (`actions.navigate()`) happens. 

This might be what we want but it should be a conscious decision.

File: packages/interactivity-router/src/index.ts [hunk 3]
```diff
@@ -103,15 +110,28 @@ const regionsToVdom: RegionsToVdom = async ( dom, { vdom } = {} ) => {
...
+		...page.scriptModules.map(
+			( src ) => import( /* webpackIgnore: true */ src )
+		),
+	] );
+	// Replace style sheets.
+	const sheets = await Promise.all( page.styles );
+	window.document
+		.querySelectorAll( 'style,link[rel=stylesheet]' )
+		.forEach( ( element ) => element.remove() );
+	window.document.adoptedStyleSheets = sheets;
```
Comment (michalczaplinski) [inline packages/interactivity-router/src/index.ts:130]: Would removing the styles and then re-adding them as `adoptedStyleSheets` not cause a flash of unstyled content?

Comment (DAreRodz) [reply packages/interactivity-router/src/index.ts:130]: As that change happens synchronously, I think there won't be any problem. 🙂

File: packages/interactivity-router/src/index.ts (hunk 1 above)
Comment (DAreRodz) [reply packages/interactivity-router/src/index.ts:197]: Uhm, I'm not sure. This `baseUrl` is required to resolve any assets on that page with relative URLs, either in the `href` attribute or in its text content, so maybe it is more robust to keep the way it is right now. 🤔

**Comment by DAreRodz** [comment]
@michalczaplinski 

> In this PR, we're effectively changing the behavior of prefetching. Previously, to "prefetch" meant to fetch the HTML **_as well as the JS and CSS assets of a page_**.
> 
> Now [...] prefetching **_only fetches the HTML_** of a new page and fetching of JS and CSS assets is delayed until a navigation (`actions.navigate()`) happens.

Oh, I see what you mean. Before, the `Promise< Page >` returned by `fetchPage` resolved when everything on the page was fully fetched, not only the HTML but also the assets. The changes in this PR also make `fetchPage` fetch the assets; it's just that it doesn't wait for them to be ready.

In practice, it would behave the same as before, as the `renderRegions` function waits for the assets to be ready before updating the HTML. However, there's a side benefit: we can now access the returned `Page` instance and know whether the client-side navigation is disabled before the assets are fully fetched, thus reloading the page without having to wait for the assets.

https://github.com/WordPress/gutenberg/blob/769342ba782fd14adf68dac2564ddbf05c84e467/packages/interactivity-router/src/index.ts#L330-L334

I have to admit this was an unconscious change. 😄 But maybe it is good to keep it this way? 🤔

**Comment by michalczaplinski** [comment]
> Oh, I see what you mean. Before, the `Promise< Page >` returned by `fetchPage` resolved when everything on the page was fully fetched, not only the HTML but also the assets.

Yup, exactly. 

Your approach is also more straightforward to implement as you can see by the number of lines you deleted 😄 

> I have to admit this was an unconscious change. 😄 But maybe it is good to keep it this way? 🤔

Maybe? 🙂 Your approach is also the one taken by Hotwired. They only prefetch the HTML and then [merge the head of previous & new pages](https://github.com/hotwired/turbo/blob/1007de974350ab37d08fa15ba04349e014ffbe32/src/core/drive/page_renderer.js#L96-L112). 

Let's first test this approach on a "real" site to see if the performance is acceptable. Previously, the new page would load instantly since everything (HTML, CSS & JS) was already loaded. Now, there might be a slight delay if we have to wait for the CSS & JS. 

If the performance is not acceptable, we add the prefetching similar to how it worked before. It worked fine for JS modules using `link` elements with `modulepreload`. For styles, we can load them like you are doing and add them to a `CSSStyleSheet` but **_only attach to the DOM once a navigation happens_**.

File: packages/interactivity-router/src/index.ts (hunk 3 above)
Comment (michalczaplinski) [reply packages/interactivity-router/src/index.ts:130]: Gotcha, thanks!

**Comment by DAreRodz** [comment]
> If the performance is not acceptable, we add the prefetching similar to how it worked before. It worked fine for JS modules using `link` elements with `modulepreload`.

The script modules logic is going to change to support dynamic import maps, so we could address this in a subsequent PR. Don't worry. 😄 

> For styles, we can load them like you are doing and add them to a `CSSStyleSheet` but **_only attach to the DOM once a navigation happens_**.

That's how it works in this PR, right? 🤔  The moment `navigate` reaches the point where the HTML is about to be updated, the `CSSStyleSheet`  instances related to the new page are applied through `document.adoptedStyleSheets`. Both the style and the HTML updating happen synchronously.

https://github.com/WordPress/gutenberg/blob/769342ba782fd14adf68dac2564ddbf05c84e467/packages/interactivity-router/src/index.ts#L130-L154

**Comment by sethrubenstein** [comment]
Hi 👋. Just dropping in to say I'm watching this and subsequent module/js loading with much excitement. Right now I'm getting around this in more complex IAPI routing applications by enqueueing on page load the modules/scripts and styles I know I'll need for new blocks that can possibly render on a new router navigation, it works, but having this be dynamic and automatic would be very welcome.

File: packages/interactivity-router/src/assets/styles.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+	if ( elementSheet ) {
+		return getCachedSheet( sheetId, () => {
+			const sheet = new CSSStyleSheet();
+			for ( const { cssText } of elementSheet.cssRules ) {
+				sheet.insertRule( withAbsoluteUrls( cssText, sheetUrl ) );
+			}
+			return Promise.resolve( sheet );
+		} );
+	}
```
Comment (luisherranz) [inline packages/interactivity-router/src/assets/styles.ts:49]: In which case the `element.sheet` (`CSSStyleSheet` instance) exists but the URLs are wrong and need to be corrected? Is `element.sheet` not only present when the `link` tag is actually on DOM, therefore only on the initial page load?

**Review by luisherranz — APPROVED** [review]
This looks fantastic. Thanks, David! Congratulations on the tests. Also, very good. 

I just have one question, but it's not limiting. Feel free to merge it so people can try it as soon as possible.

Regarding preloading of CSS and JS assets, we can discuss it and add it back in a subsequent PR.

File: packages/interactivity-router/src/index.ts (hunk 1 above)
Comment (luisherranz) [reply packages/interactivity-router/src/index.ts:197]: Yes, we don't have to change the behavior that the browser would do by default in a server-side navigation.

**Comment by michalczaplinski** [comment]
> The script modules logic is going to change to support dynamic import maps, so we could address this in a subsequent PR. Don't worry. 😄

Not worried :) A follow-up PR sounds good 👍 

> That's how it works in this PR, right? 🤔 The moment navigate reaches the point where the HTML is about to be updated, the CSSStyleSheet instances related to the new page are applied through document.adoptedStyleSheets. Both the style and the HTML updating happen synchronously.

Yes, that's right. I didn't mean to imply otherwise 🙂.  

---

I also reviewed the logic for `updateHead()` from `head.ts` that was removed in this PR. Removing it is fine for now but we should revisit that logic in subsequent PRs. Before this PR, we would replace all the `<link>` elements with the ones coming from the "new" page.

In general, some other things can be present in  `<head>` that we should handle:

- `<link rel="canonical">` elements should be updated in the head when doing a full-page navigation. In the current PR, they will still point to the previous page.
- Same for `<link rel="shortlink">`.
- `<link>` elements that load assets other than CSS like fonts, audio, etc.
- This might be an edge case, but `<head>` can also contain `<template>` elements.

**Pull request merged**